Spigot-Server-Patches/0245-Expose-client-protocol-version-and-virtual-host.patch

```diff
@@ -0,0 +1,140 @@
+From e05a60bfd0fc59eff42a4ef78d2c2889019c0cd5 Mon Sep 17 00:00:00 2001
+From: Minecrell <minecrell@minecrell.net>
+Date: Tue, 10 Oct 2017 18:45:20 +0200
+Subject: [PATCH] Expose client protocol version and virtual host
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/network/PaperNetworkClient.java b/src/main/java/com/destroystokyo/paper/network/PaperNetworkClient.java
+new file mode 100644
+index 000000000..5caca6439
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/network/PaperNetworkClient.java
+@@ -0,0 +1,50 @@
++package com.destroystokyo.paper.network;
++
++import net.minecraft.server.NetworkManager;
++
++import java.net.InetSocketAddress;
++
++import javax.annotation.Nullable;
++
++public class PaperNetworkClient implements NetworkClient {
++
++    private final NetworkManager networkManager;
++
++    PaperNetworkClient(NetworkManager networkManager) {
++        this.networkManager = networkManager;
++    }
++
++    @Override
++    public InetSocketAddress getAddress() {
++        return (InetSocketAddress) this.networkManager.getSocketAddress();
++    }
++
++    @Override
++    public int getProtocolVersion() {
++        return this.networkManager.protocolVersion;
++    }
++
++    @Nullable
++    @Override
++    public InetSocketAddress getVirtualHost() {
++        return this.networkManager.virtualHost;
++    }
++
++    public static InetSocketAddress prepareVirtualHost(String host, int port) {
```

Comment:
I assume this has been tested with BungeeCord - how about LilyPad?

Reply:
It looks like they replace the hostname entirely with a JSON string: https://github.com/LilyPad/GoLilyPad/blob/master/server/proxy/sessionOutBridge.go#L73

It shouldn't break anything, but the method won't return the correct result. I don't think we can do much here, we'd need to detect LilyPad and parse the JSON manually to get the correct result. I don't think this is in scope for Paper.

Reply:
Actually it seems like they hack the handshake packet to set the correct hostname again so it might actually work: https://github.com/LilyPad/Bukkit-Connect/blob/master/src/main/java/lilypad/bukkit/connect/login/LoginNettyInjectHandler.java#L113-L118

Reply:
I'm going to request a review from @electronicboy and @Zbob750 - let's see what they think.

Reply:
https://github.com/LilyPad/Bukkit-Connect/blob/master/src/main/java/lilypad/bukkit/connect/login/LoginPayload.java

Reply:
Just tested it and it's indeed returning the correct virtual host. Yay!